Panel migrates to use libpldm APIs

PriyangaRamasamy opened the pull request.

Lakshmi-y committed:
Migration:Use libpldm instance ID & transport APIs

Replace the current APIs with libpldm APIs for allocating instance IDs
directly.  This eliminates the need for remote D-Bus calls to the pldm
daemon.

Replace pldm transport APIs with libpldm pldm_transport APIs. This
change migrates the application off the deprecated "requester" APIs in
libpldm. Since we currently lack the infrastructure to obtain the
correct TIDs, use the EID as the TID in the EID-to-TID mapping to
maintain the current functionality.

Test:
Feb 27 19:22:32 p10bmc ibm-panel[894]: L1 : 41
Feb 27 19:22:32 p10bmc ibm-panel[894]: L2 :

Feb 27 19:22:49 p10bmc ibm-panel[894]: L1 : A1003041
Feb 27 19:22:49 p10bmc ibm-panel[894]: L2 :

Feb 27 19:22:51 p10bmc ibm-panel[894]: Sub function executed = 0

Feb 27 19:22:52 p10bmc ibm-panel[894]: Got instanceId: 0
from PLDM eid: 9
Feb 27 19:22:52 p10bmc ibm-panel[894]: Panel function 29.
Data packet sent to pldm: 80 02 39 08 00 01 01 29

Feb 27 19:22:52 p10bmc ibm-panel[894]: Panel function 29
executed successfully.

Feb 27 19:22:52 p10bmc ibm-panel[894]: L1 : 41    00
Feb 27 19:22:52 p10bmc ibm-panel[894]: L2 :

Change-Id: I5854aa022582ceae6a1843983fc90181b7e6d55f
Signed-off-by: Priyanga Ramasamy <priyanga24@in.ibm.com>

branupama commented on src/pldm_fw.cpp:
```diff
@@ -105,10 +115,50 @@ int PldmFramework::openMctpDemuxTransport()
+    if (rc)
+    {
+        std::cerr
+            << "openAfMctpTransport: Failed to setup tid to eid mapping. rc = "
```
tid and eid should be capital.

branupama commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,159 @@
+{
+    impl.afMctp = nullptr;
+    int rc = pldm_transport_af_mctp_init(&impl.afMctp);
+    if (rc)
```
Better to use (rc != 0) as rc is int type, will gives more clarity to reader.

same for below as well.

PriyangaRamasamy replied:
I don't find a value in explicitly checking for 0. 
reader should know the return type of pldm_transport_af_mctp_init to get the clarity.

branupama commented on src/pldm_fw.cpp:
```diff
@@ -59,7 +59,16 @@ int PldmFramework::openPldm()
+#elif defined(PLDM_TRANSPORT_WITH_AF_MCTP)
+    fd = openAfMctpTransport();
+#else
+    std::cerr << "open: No valid transport defined!" << std::endl;
```
No valid PLDM transport ?

PriyangaRamasamy replied:
there is nothing called as PLDM TRANSPORT.

branupama commented on src/pldm_fw.cpp:
```diff
@@ -59,7 +59,16 @@ int PldmFramework::openPldm()
+    fd = openAfMctpTransport();
+#else
+    std::cerr << "open: No valid transport defined!" << std::endl;
+    throw FunctionFailure("Required host dump action via PLDM is not allowed: "
```
does openPldm() caller handling exception ?
do we need to send -1 in case of error ? to create PEL in the caller .?

branupama commented on src/pldm_fw.cpp:
```diff
@@ -105,10 +115,50 @@ int PldmFramework::openMctpDemuxTransport()
+        std::cerr
+            << "openAfMctpTransport: Failed to init AF MCTP transport. rc = "
+            << rc << std::endl;
+        throw FunctionFailure("Failed to init AF MCTP transport");
```
does openPldm() caller handling exception ?
do we need to send -1 in case of error ? to create PEL in the caller .?

same for below as well

PriyangaRamasamy replied:
this method returns file descriptor and not rc. we cannot return rc to the caller. will add a catch block in caller  and log PEL there

branupama replied:
is this is pending ?

branupama commented:
> This change migrates the application off the deprecated "requester" APIs in
libpldm.
I dint understand this statement

branupama commented on include/pldm_fw.hpp:
```diff
@@ -21,12 +24,26 @@ class PldmFramework
+        if (rc)
+        {
+            std::cout << "pldm_instance_db_destroy failed, rc =" << rc << "\n";
+        }
```
should we set pldmInstanceIdDb to nullptr here ?

PriyangaRamasamy replied:
yes. will do that

branupama commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,109 @@
     }
-    catch (const sdbusplus::exception::SdBusError& e)
+
+    if (rc)
```
should it be check on -EAGAIN ?

PriyangaRamasamy replied:
if (rc == -EAGAIN) this is already checked at line 23 - that is to try to allocate instance id again..

when the first time we hit -EAGAIN - we attempt to retry.. but when the second time we hit -EAGAIN or any rc, we throw the error..

branupama replied:
okay

souvik1914581 commented on include/pldm_fw.hpp:
```diff
@@ -73,7 +74,13 @@ class PldmFramework
     pldm_transport* pldmTransport = nullptr;
 
-    pldm_transport_mctp_demux* mctpDemux = nullptr;
+    union TransportImpl
```
Any reason why we are using `union` here instead of `std::variant<>` ?

PriyangaRamasamy replied:
std::variant is for runtime detection of the types. here we will decide which transport layer has to be used at compile time itself. so we use union to allocate the shared memory space for both the types.. Based on the meson option provided at compile time, we will decide the transport type.

SunnySrivastava1984 replied:
I will need some more insight on runtime and compile time argument here.

jinuthomas reviewed (changes requested):
At a high level see two changes I feel not apt 
1. the return code checks , no where it tells me that a positive value is a failure.
2. the multi level throw in code , is that even required , does it help us in any way

branupama commented on include/pldm_fw.hpp:
```diff
@@ -21,12 +25,30 @@ class PldmFramework
+        int rc = pldm_instance_db_destroy(pldmInstanceIdDb);
+        if (rc)
+        {
+            if (pldmInstanceIdDb)
```
we don't need this check here and we can directly assign nullptr to pldmInstanceIdDb, as we going out of scope.

RekhaAparna01 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,162 @@
+    if (rc)
     {
-        std::cerr << e.what() << std::endl;
+        std::cerr << "Return code = " << rc << std::endl;
```
Can the error statement be: "Failed to get instance id. Return code = ". This will help to know which function failed?

PriyangaRamasamy replied:
done

PriyangaRamasamy commented:
> At a high level see two changes I feel not apt
> 
> 1. the return code checks , no where it tells me that a positive value is a failure.
> 2. the multi level throw in code , is that even required , does it help us in any way

1. These are the return codes from **libpldm APIs** which returns errno on failure and 0 on success. 
Eg: https://github.com/openbmc/libpldm/blob/8fd26555be33fadb0c0345897ddcbec859fda1b5/src/requester/instance-id.c#L4.
2. As a caller PldmFramework::sendPanelFunctionToPhyp() -  this API will know for which function number the execution failed and it will log a PEL for the specific reason which is thrown from the point of failure. Yes it helps us in logging PEL at the caller place.

RekhaAparna01 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,162 @@
+        std::cerr << "open: pldmTransport already setup!" << std::endl;
+        throw FunctionFailure("Required host dump action via PLDM is not "
+                              "allowed due to openPldm failed");
+        return fd;
```
since you are throwing exception, return is not required.

PriyangaRamasamy replied:
resolved

jinuthomas reviewed (changes requested):
still not convinced by the return , as nowhere is it mentioned that positive numbers are a failure , either check equals 0for success or negative for failure , will help if they make a change tomorrow for adding positive values for some reason.

I can live with the trow , but my question is is it that important , I am ok with having it there also

PriyangaRamasamy commented:
Changes made in [476b4f9](https://github.com/ibm-openbmc/ibm-panel/commit/476b4f9f3c421b1bbfd7cd0e3dd2f92e69c71b3a)
1. Updated the error check portion to check on negative errors (cross checked with libpldm apis)
2. Removed meson option to initiate transport layer. Currently we have only one way to initiate transport layer which is using mctp demux deamon running on BMC. Thus removed the unused AF MCTP logic to open and close transport layer.
3. Instance ID data base is not initiated at constructor.. It will get initiated on demand.

souvik1914581 commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,17 @@ class PldmFramework
+
+    /** @brief PLDM instance ID database object used to get instance IDs
+     */
+    pldm_instance_db* pldmInstanceIdDb = nullptr;
```
Why are we using raw pointers here instead of C++ smart pointers?

souvik1914581 replied:
Ok, I see this is dictated by `pldm_instance_db_init_default` API.

PriyangaRamasamy replied:
right

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+    rc = pldm_instance_id_alloc(pldmInstanceIdDb, tid, &instanceID);
+    if (rc == -EAGAIN)
+    {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
How did you arrive on this sleep time of 100ms before retry?

PriyangaRamasamy replied:
it's a random pick to retry the instance id allocation after 0.1 second

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+    pldm_instance_id_t instanceID = 0;
+
+    rc = pldm_instance_id_alloc(pldmInstanceIdDb, tid, &instanceID);
+    if (rc == -EAGAIN)
```
Why are we comparing `rc` with `-EAGAIN` (negative `EAGAIN`) ? `EAGAIN` would be some positive integer value defined in `errno.h` right?

souvik1914581 replied:
Ok, the `pldm_instance_id_alloc` API negates the error numbers before returning.

PriyangaRamasamy replied:
yes

souvik1914581 commented on include/pldm_fw.hpp:
```diff
@@ -81,12 +95,32 @@ class PldmFramework
-     * id is generated by pldm.
+     * This api returns the instance id by making a dbus call to libpldm
+     * api. Instance id is to uniquely identify a message packet.
+     * @return pldm_instance_id_t instance id.
```
Nit: `@throw` missing

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+                  << " of tid = " << static_cast<int>(tid) << " rc = " << rc
+                  << std::endl;
+    }
+
```
Should we try to destroy instance ID DB if freeing instance ID has failed?

PriyangaRamasamy replied:
yes. instance id DB is created whenever new instance ID is allocated and destroyed whenever the instance ID is destroyed

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+
+        if (pldmInstanceIdDb)
+        {
+            pldmInstanceIdDb = nullptr;
```
We should call `free(pldmInstanceIdDb)` to destroy the instance ID DB object "manually", and then do `pldmInstanceIdDb = nullptr`.

PriyangaRamasamy replied:
why we need both free and pldmInstanceIdDb = nullptr. ? 

using free() is not required as if pldmInstanceIdDB is already destroyed (or not created itself) then free() will throw undefined behaviour.
pldmInstanceIdDb = nullptr; will do the deallocation of memory. so we are good here

souvik1914581 replied:
Ok, I see `pldm_instance_db_destroy()` API does a `free()` on the `pldmInstanceIdDb` pointer.
Ideally, the API itself should have done `pldmInstanceIdDb = nullptr`.

souvik1914581 commented on include/pldm_fw.hpp:
```diff
@@ -81,12 +95,32 @@ class PldmFramework
+    int openMctpDemuxTransport();
+
+    /** @brief Close the PLDM file */
+    void pldmClose();
```
The open method is `openPldm()`, so to be consistent, this method should be named`closePldm()`

PriyangaRamasamy replied:
this doesn't impact the code quality. both gives same meaning. am ok with pldmClose

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+            std::to_string(rc));
+    }
+
+    pldmTransport = pldm_transport_mctp_demux_core(mctpDemux);
```
Missing return value check.

PriyangaRamasamy replied:
pldmTransport is the struct. if I use is_empty() to check it will always tell its not empty as it is_empty will check if struct has atleast one data member.. it doesn;t validate the value of that data member..
 
struct pldm_transport {
	const char *name;
	uint8_t version;
	pldm_requester_rc_t (*recv)(struct pldm_transport *transport,
				    pldm_tid_t *tid, void **pldm_resp_msg,
				    size_t *msg_len);
	pldm_requester_rc_t (*send)(struct pldm_transport *transport,
				    pldm_tid_t tid, const void *pldm_msg,
				    size_t msg_len);
	int (*init_pollfd)(struct pldm_transport *transport,
			   struct pollfd *pollfd);
};

PriyangaRamasamy replied:
so no need of return value check here

souvik1914581 replied:
I meant we should check for `nullptr` return value.

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+
+    pldmTransport = pldm_transport_mctp_demux_core(mctpDemux);
+
+    struct pollfd pollfd;
```
Nit: Suggest using a variable name which is different than the datatype, incase someone does a `typedef struct pollfd pollfd;` in the future, it may lead to compilation errors.

Also, as a good practice, suggest initializing `pollfd` members with some invalid values.

PriyangaRamasamy replied:
pollfd initialisation is taken care by **pldm_transport_mctp_demux_init_pollfd** API. so no need of default initialisation from panel application side as we cannot decide any default value.

this is how pollfd struct looks like.
struct pollfd {
	int fd;	       /* file descriptor */
	short events;  /* requested events */
	short revents; /* returned events */
};

PriyangaRamasamy replied:
>Nit: Suggest using a variable name which is different than the datatype, incase someone does a typedef struct pollfd pollfd; in the future, it may lead to compilation errors.

sure will consider this point. right now we don't have any conflicts with type and variable name. will keep it unchanged

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+{
+    pldm_transport_mctp_demux_destroy(mctpDemux);
+    mctpDemux = nullptr;
+    pldmTransport = nullptr;
```
Does `libpldm` take ownership of memory pointed to by `pldmTransport` and free it?

PriyangaRamasamy replied:
no. we should handle it

souvik1914581 replied:
I do not see a `free(pldmTransport)` anywhere in our code?

PriyangaRamasamy replied:
it is there in libpldm code 
https://github.com/openbmc/libpldm/blob/779e9dbdcfaf8e131b6c35edbc4e057aab52a65a/src/transport/mctp-demux.c#L273

souvik1914581 replied:
Ok. Got it.

souvik1914581 commented on include/pldm_fw.hpp:
```diff
@@ -81,12 +95,32 @@ class PldmFramework
+     * @brief setup PLDM transport for sending and receiving messages
      *
-     * @return one byte instance id.
+     * @return file descriptor
```
Some more description will be helpful.
`file descriptor` to do what?

PriyangaRamasamy replied:
it is the file descriptor of the socket established by PLDM for communication between BMC and host. will update in the file

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,140 @@
+    fd = openMctpDemuxTransport();
+    if (fd < 0)
+    {
+        auto e = errno;
```
What is the use of reading `errno` here?

I see `openMctpDemuxTransport()` throws exception incase of any failure, so the `if(fd < 0)` block will not execute.

In case`openMctpDemuxTransport()` does not encounter any internal failure, and returns `-1`,  I do not see it setting `errno` anywhere.

PriyangaRamasamy replied:
>What is the use of reading errno here?

errno is set by C library when the system call fails. so we need errno if we receive fd as -1.



>I do not see it setting errno anywhere.

what do you mean?


>I see openMctpDemuxTransport() throws exception incase of any failure, so the if(fd < 0) block will not execute.

see there are two things to consider here.. one is openMctpDemuxTransport() API has a flow to open a mctp socket.. if there is any issue in the flow then we throw exception..

if flow succeded there is a chance that we may receive filedescriptor as -1. and that is why we have these checks.. what do you mean by if(fd < 0) block will not execute?

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +220,78 @@ types::PldmPacket
-    if (fd == -1)
+        // Open PLDM socket based communication
+        int fd = openPldm();
+        if (fd == -1)
```
I see `openPldm()` throws exception incase of any errors, so this `if (fd == -1)` block is dead code.

PriyangaRamasamy replied:
I'll remove this check

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +220,78 @@ types::PldmPacket
+
+        freeInstanceID(instance);
+        pldmClose();
+        if (rc < 0)
```
Why is the return value being checked down here and not immediately after `pldm_transport_send_msg()` call?

PriyangaRamasamy replied:
because before throwing the exception saying SEND FAILED, we are doing certain things, viz., 
1. we are capturing the data  packet sent on console 
2. we are freeing up the instance ID
3. we are closing the PLDM socket.

these are required before we throw and exit

souvik1914581 replied:
Ok.

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,131 @@
+            pldmInstanceIdDb = nullptr;
+            std::cout << "\n Destroyed pldm instance ID DB object manually"
+                      << std::endl;
+        }
```
`if(rc >= 0)` case implies memory pointed to by `pldmInstanceIdDb` has been freed by `pldm_instance_db_destroy()`, the `pldmInstanceIdDb` pointer needs to be set to `nullptr` here.

PriyangaRamasamy replied:
so irrespective of rc, we need to set pldmInstanceIdDb to nullptr.

souvik1914581 replied:
Correct, because `pldmInstanceIdDb` becomes a dangling pointer after `pldm_instance_db_destroy()` and should be set to `nullptr` to prevent any future dereference.

souvik1914581 reviewed (changes requested):
`nullptr` return value check missing in one place and dead code removal pending.
Otherwise looks fine to me.

SunnySrivastava1984 commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,17 @@ class PldmFramework
     /** Host mctp eid */
     static constexpr auto mctpEid = (types::Byte)9;
 
+    static constexpr types::TerminusID tid = mctpEid;
```
A one liner to explain what it is or what it will be used for will be handy?

PriyangaRamasamy replied:
its the host terminus ID which is already there in include/types.hpp..
For more details refer PLDM and libpldm documentation.

SunnySrivastava1984 replied:
Then why is it there for "pldmInstanceIdDb", "pldmTransport" and "mctpDemux".
Can't that be referred from somewhere?

It was asked fro consistency through out the file.

PriyangaRamasamy replied:
done

SunnySrivastava1984 commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,17 @@ class PldmFramework
 
+    static constexpr types::TerminusID tid = mctpEid;
+
+    /** @brief PLDM instance ID database object used to get instance IDs
```
We don't use @brief in case of member variable. Should be removed for consistency?

PriyangaRamasamy replied:
removed

SunnySrivastava1984 commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,17 @@ class PldmFramework
+    /** pldm transport instance  */
+    pldm_transport* pldmTransport = nullptr;
+
+    pldm_transport_mctp_demux* mctpDemux = nullptr;
```
Same here, a one liner.

PriyangaRamasamy replied:
added

SunnySrivastava1984 commented on include/pldm_fw.hpp:
```diff
@@ -81,12 +95,40 @@ class PldmFramework
+     */
+    int openPldm();
+
+    /** @brief Opens the MCTP socket for sending and receiving messages.
```
@brief should be in the next line.

PriyangaRamasamy replied:
done

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,135 @@
 #include <libpldm/platform.h>
 #include <libpldm/pldm.h>
 #include <libpldm/state_set.h>
+#include <poll.h>
```
Just curious, if the meson check 
if cxx.has_header('poll.h') fails, still this will compile?

PriyangaRamasamy replied:
check fails ? did you mean if the check returns false?

if it returns false, compilation will not fail.

SunnySrivastava1984 replied:
My concern was if you are checking "if cxx.has_header('poll.h')" then you imply that there is a possibility of compiler not having poll.h.
So if it is actually missing then in that case how will "#include <poll.h>" compile? From where it will get included?

PriyangaRamasamy replied:
right, I added an ifdef check before including poll.h

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,135 @@
+    rc = pldm_instance_id_alloc(pldmInstanceIdDb, tid, &instanceID);
+    if (rc == -EAGAIN)
+    {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Not sure, how much change can happen under 100 milli seconds. Should the wait be for some considerable time? Else what is the point re-trying?

PriyangaRamasamy replied:
ideally instance id allocation should not fail.. But for some reason if libpldm fails to get instance id for it's users - it returns EAGAIN asking us to try querying again. I don't find any rule in libpldm asking us to retry after a delay. This delay is for our safer side. 100ms is working fine till now. I would suggest if we hit any issues then we can modify this delay.

SunnySrivastava1984 replied:
Ok.

jinuthomas replied:
I have a question on this based on your last sentence , since you said that this working fine , wanted to know how did you test this condition to be working fine ?

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +215,71 @@ types::PldmPacket
+        }
+
+        // Open PLDM socket based communication
+        [[maybe_unused]] int fd = openPldm();
```
Am I missing something? I don't see "fd" being used anywhere. Why are we declaring it?

PriyangaRamasamy replied:
we need it for error logging.. will update

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,135 @@
     }
-    return instanceId;
+
+    fd = openMctpDemuxTransport();
```
Could have returned the function call itself. Don't see a need of another local variable.

PriyangaRamasamy replied:
modified

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,133 @@
+
+void PldmFramework::freeInstanceID(pldm_instance_id_t instanceID)
+{
+    int rc = pldm_instance_id_free(pldmInstanceIdDb, tid, instanceID);
```
What happens if "pldmInstanceIdDb" is a null pointer?

PriyangaRamasamy replied:
pldm_instance_id_free() returns -EINVAL

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,133 @@
+        std::cout << "pldm_instance_db_destroy failed, rc =" << rc << "\n";
+    }
+
+    if (pldmInstanceIdDb)
```
Do we need both ways of destruction? Why can't we just close it manually and we are done?

PriyangaRamasamy replied:
yes we need to tell libpldm that this instance of DB got destroyed. and also we need to set null in our class member variable.

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,133 @@
+        throw FunctionFailure("Failed to init MCTP demux transport");
+    }
+
+    rc = pldm_transport_mctp_demux_map_tid(mctpDemux, tid, tid);
```
one of the tid is supposed to be "pldm tid" and another is supposed to be "mctp eid" if I am not wrong. 
Now since the value and type is same this will work.
But the variable that is being used here does not state that at all. Expect a better nomenclature here.

PriyangaRamasamy replied:
The issue is not with nomenclature. Here we intentionally send tid in the place of eid.
Why because EID is the entity ID which is a specific term used in PLDM. DMTF has no way to distinguish between TID and EID. 

PLDM will build the infrastructure to get the correct TID from the EID. Until then we use eid as tid in EID-TID mapping.

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,133 @@
+    }
+
+    rc = pldm_transport_mctp_demux_map_tid(mctpDemux, tid, tid);
+    if (rc < 0)
```
Is this check valid?
"pldm_transport_mctp_demux_map_tid" will return any negative value?

PriyangaRamasamy replied:
it returns 0 now. but its return type is int and on error cases it may return negative numbers in future when there is any change in library functions. That's why I have this check here.

SunnySrivastava1984 replied:
How is it known that in future it will always have negative value on an error case?
Unless that is guaranteed, this still is not a valid code.

PriyangaRamasamy replied:
https://github.com/openbmc/libpldm/blob/5192e2e26f26ffbbf4256429ea5ae42237e011cc/include/libpldm/pldm.h#L15
the above shows that the requester error codes are negative integers.

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,133 @@
+
+    struct pollfd pollfd;
+    rc = pldm_transport_mctp_demux_init_pollfd(pldmTransport, &pollfd);
+    if (rc < 0)
```
Can "pldm_transport_mctp_demux_init_pollfd" return a negative value? I don't see this API returning negative value.
https://github.com/openbmc/libpldm/blob/779e9dbdcfaf8e131b6c35edbc4e057aab52a65a/src/transport/mctp-demux.c#L74

Am I referring to a wrong place? If so please correct me, else
I don't see a need for return type being checked for negative value.

PriyangaRamasamy replied:
it returns 0 now. but its return type is int and on error cases it may return negative numbers in future when there is any change in library functions. That's why I have this check here.

SunnySrivastava1984 replied:
How is it known that in future it will always have negative value on an error case?
Unless that is guaranteed, this still is not a valid code.

PriyangaRamasamy replied:
https://github.com/openbmc/libpldm/blob/5192e2e26f26ffbbf4256429ea5ae42237e011cc/include/libpldm/pldm.h#L15
the above shows that the requester error codes are negative integers.

SunnySrivastava1984 commented on meson.build:
```diff
@@ -34,6 +34,10 @@ systemd_system_unit_dir = systemd.get_variable('systemdsystemunitdir')
 service_file = 'service_files/com.ibm.panel.service'
 
+if cxx.has_header('poll.h')
+  add_project_arguments('-DPLDM_HAS_POLL=1', language: 'cpp')
```
Why are we adding this? Not able to understand the use case.

PriyangaRamasamy replied:
libpldm has this config check to use polling APIs. since we are using the poll apis directly from libpldm, we need this config check

souvik1914581 commented on include/pldm_fw.hpp:
```diff
@@ -81,12 +97,41 @@ class PldmFramework
+
+    /**
+     * @brief Free PLDM requester instance id
+     * @param[in] pdrs - Panel PDR data.
```
`@param[in]` does not match with actual parameter

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -8,32 +8,151 @@
+    }
+
+    // PLDM_INST_ID_MAX is 32 (0 to 31). Initialising with an invalid/unused ID.
+    pldm_instance_id_t instanceID = 100;
```
We should initialize with `PLDM_INST_ID_MAX + 1` to make sure `instanceID` here is always invalid.

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -8,32 +8,151 @@
+    if (pldmInstanceIdDb)
+    {
+        // PLDM_INST_ID_MAX is 32 (0 to 31)
+        if (instanceID <= 31)
```
Should be `instanceID <= PLDM_INST_ID_MAX - 1`
I see `PLDM_INST_ID_MAX` in `instance-id.c` file of libpldm. We should have a macro defined in our code which is a copy of the library's macro.

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -8,32 +8,151 @@
+        // PLDM_INST_ID_MAX is 32 (0 to 31)
+        if (instanceID <= 31)
+        {
+            int rc = pldm_instance_id_free(pldmInstanceIdDb, tid, instanceID);
```
`pldm_instance_id_free` can return `-EAGAIN`. We should add re-try logic here as well.

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +232,75 @@ types::PldmPacket
-        utils::getPDR(phypTerminusID, frontPanelBoardEntityId,
-                      stateIdToEnablePanelFunc, "FindStateEffecterPDR");
+    // PLDM_INST_ID_MAX is 32 (0 to 31). Initialising with an invalid/unused ID.
+    pldm_instance_id_t instance = 100;
```
Should be initialized with `PLDM_INST_ID_MAX`.

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -8,32 +8,151 @@
+        pldm_transport_mctp_demux_destroy(mctpDemux);
+        mctpDemux = nullptr;
+    }
+
```
A comment here which says `pldm_transport_mctp_demux_destroy()` takes care of de-allocating memory pointed to by `pldmTransport` would be useful.

PriyangaRamasamy replied:
done

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -8,32 +8,151 @@
+
+void PldmFramework::freeInstanceID(pldm_instance_id_t instanceID)
+{
+    if (pldmInstanceIdDb)
```
nit: `if(!pldmInstanceIdDb) { return; } ` would be more readable.
We wouldn't need to ensure all the subsequent code is inside the `if(pldmInstanceIdDb)` block.

PriyangaRamasamy replied:
done

souvik1914581 commented on meson.build:
```diff
@@ -55,72 +64,57 @@ panel_app_a = static_library(
     'src/executor.cpp',
     'src/pldm_fw.cpp',
-    include_directories: 'include'
+    include_directories: 'include',
```
The `,` got added by `clang` ?

souvik1914581 commented on meson.build:
```diff
@@ -55,72 +64,57 @@ panel_app_a = static_library(
     'src/executor.cpp',
     'src/pldm_fw.cpp',
-    include_directories: 'include'
+    include_directories: 'include',
```
The `,` got added by `clang` ?

PriyangaRamasamy replied:
yes

souvik1914581 commented on meson.build:
```diff
@@ -55,72 +64,57 @@ panel_app_a = static_library(
-     include_directories: ['include', 'tools/include']
+    'ibm_dbus_call_a',
+    'tools/src/dbus_call.cpp',
+    include_directories: ['include', 'tools/include'],
```
The `,` got added by `clang` ?

PriyangaRamasamy replied:
yes

souvik1914581 reviewed (commented):
Few minor comments, other than that looks good to me.

SunnySrivastava1984 commented on meson.build:
```diff
@@ -20,29 +20,38 @@ phosphor_dbus_interfaces = dependency('phosphor-dbus-interfaces')
-]),
-language : 'cpp')
-add_global_arguments('-Wno-psabi', language : ['c', 'cpp'])
+    cxx.get_supported_arguments(
```
Why are we taking clang formatting of meson file via this PR?
Shouldn't this be a separate PR to fix clang? This change unnecessarily makes both the changes dependent.

PriyangaRamasamy replied:
anyways this changes are dependent on clang fix.. that's why I pushed in the same PR.

PriyangaRamasamy replied:
but with a different commit.

PriyangaRamasamy replied:
moved to separate commit

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -8,32 +8,161 @@
+
+    // Destroy the instance ID DB
+    int rc = pldm_instance_db_destroy(pldmInstanceIdDb);
+    if (rc < 0)
```
Why is this being checked for less than 0? Called API will always return 0.

PriyangaRamasamy replied:
On failure libpldm apis will return value less than 0.

The below link has the requester's expected return value on success and failure. https://github.com/openbmc/libpldm/blob/5192e2e26f26ffbbf4256429ea5ae42237e011cc/include/libpldm/pldm.h#L15

SunnySrivastava1984 commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +242,75 @@ types::PldmPacket
+        }
+
+        // Open PLDM socket based communication
+        int fd = openPldm();
```
Not sure if we should continue if the fd is "0" as the "openMctpDemuxTransport" failed? or is it ok to continue?

PriyangaRamasamy replied:
libpldm returns fd as 0 when PLDM_HAS_POLL config is not declared. So keeping panel in sync with libpldm.

https://github.com/openbmc/libpldm/blob/5192e2e26f26ffbbf4256429ea5ae42237e011cc/src/transport/transport.c#L27C19-L27C23

souvik1914581 commented on include/const.hpp:
```diff
@@ -123,5 +123,8 @@ static const uint8_t LEN_OF_RAW_HEX_WORD = 4;
 static const std::string defaultHexWordValue = "00000000";
+
+// Maximum valid instance IDs' which can be created by libpldm
+static const auto PLDM_INST_ID_MAX = 32;
```
nit: this can be `constexpr`

souvik1914581 replied:
nit: this can be `constexpr`

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,154 @@
+            std::to_string(rc));
+    }
+
+    std::cout << "Got instanceId: " << static_cast<int>(instanceID)
```
Do these traces add any value or are they just there for debugging purpose?

PriyangaRamasamy replied:
it's there for debugging purpose

souvik1914581 replied:
Alright. Ok with this.

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,154 @@
+        }
+    }
+
+    // Destroy the instance ID DB
```
Couple of points here:
1. If `instanceID >= PLDM_INST_ID_MAX`, should we proceed to destroy the instance ID Database?
2.  The doxygen of `freeInstanceID` tells `Free PLDM requester instance id`, but it is also destroying the instance ID Database, which is another functionality. I think "destroying the instance ID Database" should be moved to another method.

PriyangaRamasamy replied:
1.  I can add a check if (pldmInstanceIdDb != nullptr) before destroying.
2. the idea here is to have single function to allocate and deallocate instance id and instance id DB. so that the caller won't miss freeing up instance ID DB. I can update the doxygen and api name..

souvik1914581 replied:
As discussed, I am fine with this.

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,154 @@
+              << std::endl;
+}
+
+int PldmFramework::openPldm()
```
The only logic inside `openPldm()` currently is to check if pldmTransport is already setup and then it calls `openMctpDemuxTransport()`.

Can we move "check if pldmTransport is already setup" inside `openMctpDemuxTransport()` and eliminate `openPldm()` ?

PriyangaRamasamy replied:
in future we will have to get rid of openMctpDemuxTransport and move to AF Mctp Transport layer 
ref: https://github.com/ibm-openbmc/ibm-panel/pull/213/files#diff-42320549c8b243f96275cd488c49a08636a92e9c8c7b178351fd5d75911f1091R66

so let's have a wrapper openPldm as a common API - which can chose between either mctp demux or af mctp demux or any other layer for opening the communication socket.

souvik1914581 replied:
Ok, I am fine with this.

souvik1914581 commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +234,75 @@ types::PldmPacket
+        }
+        std::cout << std::endl;
+
+        if (rc != 0)
```
This `rc` should be compared with `PLDM_REQUESTER_SUCCESS` incase the `pldm_requester_rc_t` enum values in the library change in future. 

Also, the doxygen of `pldm_transport_send_msg` mentions `errno` maybe set. The `errno` should also be captured as part of the `exception.what()`

PriyangaRamasamy replied:
done

souvik1914581 reviewed (approved):
LGTM

jinuthomas commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,19 @@ class PldmFramework
     static constexpr auto mctpEid = (types::Byte)9;
 
+    /** Terminus ID to which the packet has to be sent */
+    static constexpr types::TerminusID tid = mctpEid;
```
why is tid needed ?

PriyangaRamasamy replied:
terminus id is the pldm terminology of pointing the destination at which the packet has to be sent.
eid is the mctp id of the host.. 
https://github.com/openbmc/pldm/blob/4bf3ed840f13c31720ffa7c255942f3836402f4f/common/transport.cpp#L18

jinuthomas commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,19 @@ class PldmFramework
     /** Host mctp eid */
     static constexpr auto mctpEid = (types::Byte)9;
```
can you rename this to hostEid

jinuthomas replied:
Also which is the host here hosboot or PHYP ?

PriyangaRamasamy replied:
host here is phyp.

jinuthomas commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,19 @@ class PldmFramework
+    static constexpr types::TerminusID tid = mctpEid;
+
+    /** PLDM instance ID database object used to get instance IDs */
+    pldm_instance_db* pldmInstanceIdDb = nullptr;
```
nit: pldmInstanceDb

jinuthomas commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,19 @@ class PldmFramework
+    /** Terminus ID to which the packet has to be sent */
+    static constexpr types::TerminusID tid = mctpEid;
+
+    /** PLDM instance ID database object used to get instance IDs */
```
instance database object not instance ID database object

jinuthomas commented on include/pldm_fw.hpp:
```diff
@@ -43,6 +46,19 @@ class PldmFramework
+
+    /** PLDM MCTP demux object which is required to establish MCTP socket based
+     * communication. */
+    pldm_transport_mctp_demux* mctpDemux = nullptr;
```
should this not be named pldmSocket ?

jinuthomas replied:
would be better to have it generic so we do not change later

jinuthomas commented on include/pldm_fw.hpp:
```diff
@@ -81,12 +97,46 @@ class PldmFramework
+     * @brief Opens the MCTP socket for sending and receiving messages.
      *
-     * @return one byte instance id.
+     * @return file descriptor to perform MCTP demux based communication.
```
just file descriptor of socket would be good

PriyangaRamasamy replied:
done

jinuthomas reviewed (changes requested):
I do understand that you have used the variables similar to what it was earlier , Also I do understand that you coded looking at the PLDM header, I would like to suggest some extra changes , if you do find that not good, maybe I can do the changes on top of your changes. I am sorry but I felt a bit confused and thought it would be better to have a discussion with you and the team to see if we can generalize this more and save us from knowing what PLDM does internally from the code.

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,143 @@
 {
-    types::Byte instanceId = 0;
-    auto bus = sdbusplus::bus::new_default();
+    // Fetch Instance ID DB
```
nit: // Fetch instance DB

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,143 @@
+    rc = pldm_instance_id_alloc(pldmInstanceDb, tid, &instanceID);
+    if (rc == -EAGAIN)
+    {
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
not sure if this is even valid .
How did you test that 100 milliseconds is a good value ?
If cannot be tested , do recommend to remove this code. 
What does PLDM library tell if this fails ?

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,143 @@
+            std::to_string(rc));
+    }
+
+    return instanceID;
```
Do we need to check if the value is valid i.e within the MAX alloted value

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,143 @@
+    }
+
+    // Initialising with an invalid/unused ID.
+    pldm_instance_id_t instanceID = constants::PLDM_INST_ID_MAX + 1;
```
why a +1 here as per your later comment in code I see the below
// constants::PLDM_INST_ID_MAX is 32 (0 to 31)

jinuthomas commented on include/pldm_fw.hpp:
```diff
@@ -41,7 +44,20 @@ class PldmFramework
     // use PLDM defined header file to refer following constants.
     /** Host mctp eid */
-    static constexpr auto mctpEid = (types::Byte)9;
+    static constexpr auto hostEid = (types::Byte)9;
```
why do you need this variable ?

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +223,76 @@ types::PldmPacket
 
-    int fd = pldm_open();
-    if (fd == -1)
+        std::cout << "Panel function " << static_cast<int>(funcNumber)
```
move this to the failure path

jinuthomas commented on include/types.hpp:
```diff
@@ -85,6 +85,9 @@ using PendingAttributesItemType =
 using PendingAttributesType = std::vector<PendingAttributesItemType>;
 
+/** @brief Host terminus ID. */
+using TerminusID = uint8_t;
```
why is this needed ?

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,143 @@
+    {
+        int rc = pldm_instance_id_free(pldmInstanceDb, tid, instanceID);
+
+        if (rc == -EAGAIN)
```
same question as in alloc 
is this even needed and how do we justify the 100 ms

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,143 @@
+
+        if (rc != 0)
+        {
+            std::cerr << "pldm_instance_id_free failed to free id = "
```
Why a print here and no error log ?

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -113,77 +223,76 @@ types::PldmPacket
-        utils::getPDR(phypTerminusID, frontPanelBoardEntityId,
-                      stateIdToEnablePanelFunc, "FindStateEffecterPDR");
+    // Initialising with invalid/unused ID
+    pldm_instance_id_t instance = constants::PLDM_INST_ID_MAX + 1;
```
why the +1 ?

jinuthomas commented on src/pldm_fw.cpp:
```diff
@@ -7,33 +7,143 @@
+        }
+    }
+
+    // Destroy the instance ID DB
```
nit: instance DB

jinuthomas reviewed (commented):
have a bit of concerns in the error path
